daemons/attrd/attrd_cib.c

```diff
@@ -153,41 +153,44 @@ static void
 attrd_erase_cb(xmlNode *msg, int call_id, int rc, xmlNode *output,
                void *user_data)
 {
-    do_crm_log_unlikely(((rc != pcmk_ok)? LOG_NOTICE : LOG_DEBUG),
-                        "Cleared transient attributes: %s "
-                        CRM_XS " xpath=%s rc=%d",
-                        pcmk_strerror(rc), (char *) user_data, rc);
+    const char *node = pcmk__s((const char *) user_data, "a node");
+
+    if (rc == pcmk_ok) {
+        crm_info("Cleared transient node attributes for %s from CIB", node);
+    } else {
+        crm_err("Unable to clear transient node attributes for %s from CIB: %s",
+                (const char *) node, pcmk_strerror(rc));
+    }
 }
 
 #define XPATH_TRANSIENT "//node_state[@uname='%s']/" XML_TAG_TRANSIENT_NODEATTRS
 
 /*!
  * \internal
- * \brief Wipe all transient attributes for this node from the CIB
+ * \brief Wipe all transient node attributes for a node from the CIB
  *
- * Clear any previous transient node attributes from the CIB. This is
- * normally done by the DC's controller when this node leaves the cluster, but
- * this handles the case where the node restarted so quickly that the
- * cluster layer didn't notice.
- *
- * \todo If pacemaker-attrd respawns after crashing (see PCMK_ENV_RESPAWNED),
- *       ideally we'd skip this and sync our attributes from the writer.
- *       However, currently we reject any values for us that the writer has, in
- *       attrd_peer_update().
+ * \param[in] node  Node to clear attributes for
  */
-static void
-attrd_erase_attrs(void)
+void
+attrd_cib_erase_transient(const char *node)
 {
     int call_id = 0;
-    char *xpath = crm_strdup_printf(XPATH_TRANSIENT, attrd_cluster->uname);
+    char *xpath = NULL;
+
+    CRM_CHECK(node != NULL, return);
+
+    xpath = crm_strdup_printf(XPATH_TRANSIENT, node);
 
-    crm_info("Clearing transient attributes from CIB " CRM_XS " xpath=%s",
-             xpath);
+    crm_debug("Clearing transient node attributes for %s from CIB using %s",
+              node, xpath);
 
     call_id = the_cib->cmds->remove(the_cib, xpath, NULL, cib_xpath);
-    the_cib->cmds->register_callback_full(the_cib, call_id, 120, FALSE, xpath,
-                                          "attrd_erase_cb", attrd_erase_cb,
-                                          free);
+    free(xpath);
+
+    // strdup() is just for logging here, so ignore failure
+    the_cib->cmds->register_callback_full(the_cib, call_id, 120, FALSE,
+                                          strdup(node), "attrd_erase_cb",
+                                          attrd_erase_cb, free);
```

Comment:
Just curious... could you pass `node` instead of `strdup(node)` and then ditch the free argument at the end?

Reply:
No, node may no longer be valid when the callback is called